lib/fs/fs_resize.cpp

```diff
@@ -0,0 +1,141 @@
+#include "fs.hpp"
+
+int Fs::allocate_datablock(INode_Data *inode_data, u_int64_t *datablock_num) {
+  int result;
+
+  for (size_t i = 0; i < NUMBER_OF_DIRECT_BLOCKS; ++i) {
+    result =
+        allocate_indirect(&(inode_data->direct_blocks[i]), 0, datablock_num);
+    if (result <= 0)
+      return result;
+  }
+
+  result =
+      allocate_indirect(&(inode_data->single_indirect_block), 1, datablock_num);
+  if (result <= 0)
+    return result;
+
+  result =
+      allocate_indirect(&(inode_data->double_indirect_block), 2, datablock_num);
+  if (result <= 0)
+    return result;
+
+  result =
+      allocate_indirect(&(inode_data->triple_indirect_block), 3, datablock_num);
+  if (result <= 0)
+    return result;
+
+  return -1;
+}
+
+// This can simply be made non recursive by copy pasting - it is just written
+// this way as a proof of concept
+int Fs::allocate_indirect(u_int64_t *storage, int n, u_int64_t *datablock_num) {
+  char buf[IO_BLOCK_SIZE];
+  int result;
+
+  if ((*storage) == 0) {
+    if ((result = datablock_manager->new_datablock(storage)) < 0)
+      return result;
+    if (n == 0) {
+      (*datablock_num) = (*storage);
+      return 0;
+    }
+  }
+
+  if (n == 0)
+    return 1;

```

Comment:
the <=0, 0, 1 status code in this part a bit too inconsist to me and we may need to discuss on uniting them

Reply:
My logic was that there are three possible results of this function:
Terminal:
return < 0 - an error has occured
return = 0 - success a datablock has been allocated
Non-Terminal:
return > 0 - unsuccessful, there was no room to allocate a datablock

Reply:
This is important and we may put it in the comments of this file